### PR TITLE
time: Do not overflow to signal value

### PR DESCRIPTION
**Pull request by Erk-**
## Motivation

As described in #5183, Tokio will panic if you sleep far, but not too far into the future.
This can cause issues and crashes in some cases where you don't expect it which is the main motivation for solving it.

## Solution

The issue is that the time value will overflow and then get set to `u64::MAX`, `u64::MAX` is also used as the signal value for the state `STATE_DEREGISTERED` which will cause the [`when`] function to think that the time already have fired and thus been deregistered.

If the program is run in debug mode it will fail a bit earlier because of a [`debug_assert`]

## Issue with this PR

This PR will currently never complete CI since the test I have added fails to ever complete. I am not sure how that should be resolved so I have opened this PR to discuss it as well. Should it be commented out or can it be made to work instead.

[`when`]: https://github.com/tokio-rs/tokio/blob/93bde0870fd706afbca795b94ce6e46d1f878edb/tokio/src/runtime/time/entry.rs#LL125C2-L134C1
[`debug_assert`]: https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/time/entry.rs#L239

### CODE REVIEW

File: tokio/tests/time_sleep.rs [hunk 1]
```diff
@@ -267,6 +267,14 @@ async fn exactly_max() {
     time::sleep(ms(MAX_DURATION)).await;
 }
 
+#[tokio::test]
+async fn less_than_max() {
+    time::pause();
+
+    let big = std::time::Duration::from_secs(u64::MAX / 10);
+    time::sleep(big).await;
```
[the remark this answers is not in the document]
Comment (Erk-) [reply tokio/tests/time_sleep.rs:275]: That works, panics before and works afterwards, thanks

**Review by Simon-Laux — APPROVED** [review]

File: tokio/tests/time_sleep.rs [hunk 2]
```diff
@@ -267,6 +267,18 @@ async fn exactly_max() {
...
+#[tokio::test]
+async fn issue_5183() {
+    time::pause();
+
+    let big = std::time::Duration::from_secs(u64::MAX / 10);
+    // This is a workaround since awaiting sleep(big) will never finish.
+    tokio::select! {
+        _ = tokio::time::sleep(big) => {}
+        _ = tokio::time::sleep(std::time::Duration::from_nanos(1)) => {}
+    }
```
Comment (Darksonn) [inline tokio/tests/time_sleep.rs:279]: We probably want to always poll the first future first.
```suggestion
    tokio::select! {
        biased;
        _ = tokio::time::sleep(big) => {}
        _ = tokio::time::sleep(std::time::Duration::from_nanos(1)) => {}
    }
```

File: tokio/tests/time_sleep.rs [hunk 3]
```diff
@@ -267,6 +267,19 @@ async fn exactly_max() {
...
 }
 
+#[tokio::test]
+async fn issue_5183() {
+    time::pause();
+
+    let big = std::time::Duration::from_secs(u64::MAX / 10);
+    // This is a workaround since awaiting sleep(big) will never finish.
+    tokio::select! {
+    biased;
```
Comment (Darksonn) [inline tokio/tests/time_sleep.rs:277]:
```suggestion
        biased;
```

Comment (Erk-) [reply tokio/tests/time_sleep.rs]: I had it like that at first, but rustfmt disagree with it, so I have skipped the expression now.

**Review by Darksonn — APPROVED** [review]
Thanks.

File: tokio/tests/time_sleep.rs (hunk 3 above)
Comment (Darksonn) [reply tokio/tests/time_sleep.rs]: Huh? Rustfmt usually does not format macros. Can you try and remove the skip annotation to see if our CI fails it?

Comment (Erk-) [reply tokio/tests/time_sleep.rs]: I did hence the force push above, here is the specific run of the CI
https://github.com/tokio-rs/tokio/actions/runs/5229606790/jobs/9442690852

Comment (Darksonn) [reply tokio/tests/time_sleep.rs]: Weird. Oh well.

**Comment by Hocuri** [comment]
Great, thanks a huge lot @Erk- !

**Pull request merged**